server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -701,19 +701,40 @@ public String[] indexAliases(
         IndexAbstraction ia = state.metadata().getIndicesLookup().get(index);
         DataStream dataStream = ia.getParentDataStream();
         if (dataStream != null) {
+            if (skipIdentity == false && resolvedExpressions.contains(dataStream.getName())) {
+                return null;
+            }
             Map<String, DataStreamAlias> dataStreamAliases = state.metadata().dataStreamAliases();
-            Stream<DataStreamAlias> stream;
+            List<DataStreamAlias> aliasesForDataStream;
             if (iterateIndexAliases(dataStreamAliases.size(), resolvedExpressions.size())) {
-                stream = dataStreamAliases.values()
+                aliasesForDataStream = dataStreamAliases.values()
                     .stream()
-                    .filter(dataStreamAlias -> resolvedExpressions.contains(dataStreamAlias.getName()));
+                    .filter(dataStreamAlias -> resolvedExpressions.contains(dataStreamAlias.getName()))
+                    .filter(dataStreamAlias -> dataStreamAlias.getDataStreams().contains(dataStream.getName()))
+                    .toList();
             } else {
-                stream = resolvedExpressions.stream().map(dataStreamAliases::get).filter(Objects::nonNull);
+                aliasesForDataStream = resolvedExpressions.stream()
+                    .map(dataStreamAliases::get)
+                    .filter(dataStreamAlias -> dataStreamAlias != null && dataStreamAlias.getDataStreams().contains(dataStream.getName()))
+                    .toList();
+            }
+
+            List<String> requiredAliases = null;
+            for (DataStreamAlias dataStreamAlias : aliasesForDataStream) {
+                if (requiredDataStreamAlias.test(dataStreamAlias)) {
+                    if (requiredAliases == null) {
+                        requiredAliases = new ArrayList<>(aliasesForDataStream.size());
+                    }
+                    requiredAliases.add(dataStreamAlias.getName());
+                } else {
+                    // we have a non-required alias for this data stream so no need to check further
+                    return null;
```

Comment:
Why does the predicate not passing for one alias name mean that no aliases are returned? What if `(dsa) -> dsa.getName().contains("foo")` was passed in as the `Predicate<DataStreamAlias> requiredDataStreamAlias` predicate, shouldn't this still go through all of the resolved aliases instead of short-circuiting?

Reply:
I believe the `required` terminology is a bit confusing here, so I totally understand where you're coming from. 

However, I think the documentation for the method clarifies why `null` is returned
```
Only aliases where the given predicate tests successfully are returned. If the indices list contains a non-required reference to the index itself - null is returned. Returns {@code null} if no filtering is required.
```

I believe the method could use more terminology that references alias filtering as that's what it does. It's currently called in 3 places - in 2 places where we don't care about filtering at all (we just want all the aliases) where the predicates are 
```
aliasMetadata -> true,
dataStreamAlias -> true,
```
and a 3rd place where we are interested in the filtering required where the predicates are 
```
AliasMetadata::filteringRequired, 
DataStreamAlias::filteringRequired
```